File: Golang_src/medium/containerWithMostWater_11.go
package medium

// Use two pointers:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 24 ms, faster than 31.52%
// Memory Usage: 6.1 MB, less than 7.25%
func maxArea(height []int) int {
	l, r := 0, len(height)-1
	result := 0

	for l < r {
		if height[l] > height[r] {
			result = max(result, height[r]*(r-l))
			r--
		} else {
			result = max(result, height[l]*(r-l))
			l++
		}
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: Cpp_src/medium/container_with_most_water_11.cc
#include <iostream>
#include <vector>
using namespace std;

// Use two pointers
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 16 ms, faster than 95.98%
// Memory Usage: 9.9 MB, less than 29.90%
class Solution
{
public:
    int maxArea(vector<int> &height)
    {
        int result = 0;
        int index_left = 0, index_right = height.size() - 1;
        while (index_left < index_right)
        {
            result = max(result, min(height[index_left], height[index_right]) * (index_right - index_left));
            if (height[index_left] < height[index_right])
            {
                index_left++;
            }
            else
            {
                index_right--;
            }
        }

        return result;
    }
};